Make LTC codegen customizable enough for XLA migration

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76180

Provides string variables to let xla customize the generated code sufficiently enough to facilitate their migration onto LTC.

Some/all of these custom variables are expected to be short-lived for migration and eventually revert to using the original content that points to LTC functionality.

Differential Revision: [D35861778](https://our.internmc.facebook.com/intern/diff/D35861778)

wconstab committed:
WIP address XLA migration needs

[ghstack-poisoned]

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -230,6 +236,13 @@ class GenLazyNativeFuncDefinition:
+    try_get_tensor: str = "TryGetLtcTensor"
+    metrics_counter: str = 'TORCH_LAZY_FN_COUNTER("lazy::")'
+    create_tensor: str = "LazyTensor::Create"
+    create_from_first_tensor: bool = False
```
I think this is ok as a short-term crutch (and bundling them all here with their defaults makes it easy to know what to kill later). Do you think these will all go away in the next few months? Or do you think some of them will be permanent differences between XLA / TS (for stuff like metrics counter, maybe? idk).

And quick related question - how does XLA actually plumb this info into the codegen? (do these values go in the yaml?)

wconstab replied:
I think all of this is intended to go away after the migration completes, but possibly some of it would stay.

The plumbing part is part of what I was wanting to clean up.  By default i could plumb these all individually as arguments to the run_gen_lazy_tensor function, but it will be a lot of arguments.

JackCaoG replied:
I think most (if not all) of these will go away after we migrate to the `LazyTensor` class in a couple months.

wconstab committed:
Update on "WIP address XLA migration needs"




[ghstack-poisoned]

bdhirsh reviewed (approved):
Looks ok to me!

JackCaoG reviewed (approved):
Thanks!

wconstab committed:
Update on "Make LTC codegen customizable enough for XLA migration"


Provides string variables to let xla customize the generated code sufficiently enough to facilitate their migration onto LTC.

Some/all of these custom variables are expected to be short-lived for migration and eventually revert to using the original content that points to LTC functionality.

[ghstack-poisoned]

wconstab committed:
Update on "Make LTC codegen customizable enough for XLA migration"


Provides string variables to let xla customize the generated code sufficiently enough to facilitate their migration onto LTC.

Some/all of these custom variables are expected to be short-lived for migration and eventually revert to using the original content that points to LTC functionality.

[ghstack-poisoned]

wconstab commented:
@wconstab has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35861778).

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -122,6 +122,11 @@ def __call__(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+    def node_base_ctor_call(self, schema: LazyIrSchema) -> str:
+        # backends can customize the way the node base class constructor is called,
+        # as long as all of its arguments can be generated from information available from the schema
+        return f"{self.node_base}(torch::lazy::OpKind({aten_symbol(schema)})"
```
@JackCaoG the way i was thinking you would handle your xla shape computation function is as follows
1. you'd provide a new class that overrides GenLazyIr (GenXLALazyIr) 
2. it would override node_base_ctor_call
  - the schema argument provided to node_base_ctor_call gives you info about all the function arguments
  - you can directly invoke `node_ctor_inputs(schema)` if you want the exact args string that ias passed to node IR constructor; 
  https://github.com/pytorch/pytorch/blob/40f3e85005b60dab26b89c6b09e3fb7cf686d2eb/torchgen/dest/lazy_ir.py#L67
  
   - or you can use an approach like our shape inference/meta kernel
  `', '.join(str(a.name) for a in schema.filtered_args())`
  
  ping me on slack if there is something tricky and hopefully i can help

JackCaoG replied:
I can overwrite `node_base_ctor_call` but it seems like it is being called in `gen` and the way it worked today is
```
{self.node_base_ctor_call(schema)},
{{{base_ctor_value_args}}}, std::move(shapes),
....
```

If I were to overwrite `node_base_ctor_call` to include the `xla::shape` related info, I would have to make our `XlaNode` constructor looks like
```
Node::Node(torch::lazy::OpKind op, 
                      const std::function<xla::Shape()>& xla_shape_fn
                     OpList operands,
                     const std::function<torch::lazy::Shape()>& shape_fn
```

which is a bit awakward since `xla_shpae_fn` or `xla_shape` would come before the operands.. I guess it is no the end of the world but wondering if there is an easy way for me to change the order of arguments.

JackCaoG replied:
on a second thought, it might not worth all the new changes, I will change pt/xla constructor for now.

wconstab replied:
oh, i confused myself when I made that change.  My intention was that `node_base_ctor_call` function produces the whole call to the node base ctor.  I must have thought that the other arguments ({{{base_ctor_value_args}}}, std::move(shapes)) are initializer list rather than base class ctor args.  

So, you're right that you can change your ctor and make this work... but I would be happy to fix it instead to do what I originally intended.  basically, i would just put all of the stuff into `node_base_ctor_call` and then you could override it as you like.  Should I proceed with that?

If so, I can bundle that together with other changes you asked for..

JackCaoG replied:
make `node_base_ctor_call` produce the whole call to the node base ctor would be very nice. If you could bundle this with couple other issues I raised I think I am good to do(for `abs` at least).

wconstab replied:
https://github.com/pytorch/pytorch/pull/76458
